### PR TITLE
Network workflow changed with cuTensornet support

### PR DESCRIPTION
**Pull request by j9263178**
Part of this PR address https://github.com/Cytnx-dev/Cytnx/issues/213 and https://github.com/Cytnx-dev/Cytnx/issues/211.

**API changed**
Changed and added APIs for regular network:

`Launch()`: 
Remove the optimal and order argument, now it simply excecutes the contraction using the existing order.

`setOrder(bool optimal, string order)`: 
when called with `optimal=True` will find an optimal order and store it, when the  `optimal=False` and `order` is provided properly will store that user specified order.

`getOrder()`: 
Return the current existing order in Network.

**cuTensornet support**
Now when the UniTensors in a netwrok are on cuda device (require to be on the same device for now), the Network `setOrder(...)` (with  `optimal=True`) and `Launch()` will find the optimal path and execute the contraction using cuTensornet.

For now the cuTensornet APIs and variables are wrapped by a class `cutensornet` (see utils/cutensornet.hpp), Network will initialize and store a instance of it upon created.

### CODE REVIEW

File: src/utils/cutensornet.hpp
Comment (kaihsin) [inline src/utils/cutensornet.hpp:1]: the header guard is not properly protected with all the possible combination of cmake options. 
if UNI_GPU=false, all CUQUANTUM/CUTENSORNET should be disable and roll back to default!!

File: tests/Tensor_test.cpp
Comment (kaihsin) [inline tests/Tensor_test.cpp:1]: incomplete dependency on #229?

**Review by kaihsin — CHANGES_REQUESTED** [review]

**Comment by j9263178** [comment]
This PR is depend on https://github.com/Cytnx-dev/Cytnx/pull/229.

**Comment by j9263178** [comment]
Start from https://github.com/Cytnx-dev/Cytnx/pull/233/commits/a08278c2716bd810783c948bef39f40f9c2c3821 the macro guards `#ifdef UNI_GPU` are added outside `#ifdef UNI_CUQUANTUM`.

File: include/Network.hpp [hunk 1]
```diff
@@ -49,6 +52,16 @@ namespace cytnx {
...
 
+    // order line
+    std::string order_line = "";
+
+    // maintan tout leg position : (tesnor id, leg idx) for each open leg.
+    std::vector<std::pair<int, int>> TOUT_pos;
+
+#ifdef UNI_CUQUANTUM
+    cutensornet cutn;
+#endif
```
[the remark this answers is not in the document]
Comment (j9263178) [reply include/Network.hpp]: Note that currently (https://github.com/Cytnx-dev/Cytnx/pull/233/commits/a08278c2716bd810783c948bef39f40f9c2c3821) user can perform network optimization/contraction on GPU only when `USE_CUDA=on` and `USE_CUQUANTUM=on`.

**Comment by j9263178** [comment]
Start from https://github.com/Cytnx-dev/Cytnx/pull/233/commits/a34945827098cf64bc0d785af6ee63bf0717ab38 NetworkTest.Network_dense will give SegFault.

File: include/utils/cutensornet.hpp
[the remark this answers is not in the document]
Comment (j9263178) [reply include/utils/cutensornet.hpp:1]: Sorry it is not trivial to me what's the best way to hide this class from user, we want to declare a `cutensornet` object for network in `include/Network.hpp`, that means we have to keep the `include/utils/cutensornet.hpp`.

**Comment by j9263178** [comment]
There are heap buffer overflow in the .FromString() step, running `NetworkTest.Network_dense_FromString` gives output:
```
==1031868==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x613000019300 at pc 0x556fc6a5e521 bp 0x7fffa9e28750 sp 0x7fffa9e28748
READ of size 8 at 0x613000019300 thread T0
    #0 0x556fc6a5e520 in std::_Vector_base<cytnx::UniTensor, std::allocator<cytnx::UniTensor> >::~_Vector_base() /home/j9263178/anaconda3/envs/cytnx_cuda/x86_64-conda-linux-gnu/include/c++/11.3.0/bits/stl_vector.h:336
    #1 0x556fc6a5e520 in std::vector<cytnx::UniTensor, std::allocator<cytnx::UniTensor> >::~vector() /home/j9263178/anaconda3/envs/cytnx_cuda/x86_64-conda-linux-gnu/include/c++/11.3.0/bits/stl_vector.h:683
    #2 0x556fc6a5e520 in cytnx::cutensornet::~cutensornet() /home/j9263178/Cytnx/include/utils/cutensornet.hpp:21
    #3 0x556fc6a5e621 in cytnx::RegularNetwork::~RegularNetwork() /home/j9263178/Cytnx/include/Network.hpp:181
    #4 0x556fc6a5e621 in cytnx::RegularNetwork::~RegularNetwork() /home/j9263178/Cytnx/include/Network.hpp:181
...
```
which is related to the destruction of member `std::vector<UniTensor> tns` in cutensornet class, I'm not sure why this is happening.

**Comment by j9263178** [comment]
There are `new-delete-type-mismatch` in the `.FromString()` step and will cause the correspoding tests failed, running `NetworkTest.Network_dense_FromString` gives output:
```
==1170185==ERROR: AddressSanitizer: new-delete-type-mismatch on 0x6140000c6c40 in thread T0:
  object passed to delete has wrong type:
  size of the allocated type:   424 bytes;
  size of the deallocated type: 440 bytes.
    #0 0x7f0eec4c0d3f in operator delete(void*, unsigned long) ../../../../libsanitizer/asan/asan_new_delete.cpp:172
    #1 0x5627fcf89156 in cytnx::Network_base::~Network_base() /home/j9263178/Cytnx/include/Network.hpp:138
    #2 0x5627fcb225e4 in void boost::checked_delete<cytnx::Network_base const>(cytnx::Network_base const*) /home/j9263178/anaconda3/envs/cytnx_cuda/include/boost/core/checked_delete.hpp:36
    #3 0x5627fcb225e4 in cytnx::intrusive_ptr_release(cytnx::intrusive_ptr_base<cytnx::Network_base> const*) /home/j9263178/Cytnx/include/intrusive_ptr_base.hpp:47
    #4 0x5627fcb225e4 in boost::intrusive_ptr<cytnx::Network_base>::~intrusive_ptr() /home/j9263178/anaconda3/envs/cytnx_cuda/include/boost/smart_ptr/intrusive_ptr.hpp:98
    #5 0x5627fcb225e4 in boost::intrusive_ptr<cytnx::Network_base>::operator=(boost::intrusive_ptr<cytnx::Network_base> const&) /home/j9263178/anaconda3/envs/cytnx_cuda/include/boost/smart_ptr/intrusive_ptr.hpp:154
    #6 0x5627fcb225e4 in cytnx::Network::FromString(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, int const&) /home/j9263178/Cytnx/include/Network.hpp:358
    #7 0x5627fcb30ad0 in NetworkTest_Network_dense_FromString_Test::TestBody() /home/j9263178/Cytnx/tests/Network_test.cpp:23
    #8 0x5627fcf348ce in void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /home/j9263178/Cytnx/build/_deps/googletest-src/googletest/src/gtest.cc:2589
    #9 0x5627fcf348ce in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /home/j9263178/Cytnx/build/_deps/googletest-src/googletest/src/gtest.cc:2625
    #10 0x5627fcf36161 in testing::Test::Run() /home/j9263178/Cytnx/build/_deps/googletest-src/googletest/src/gtest.cc:2664
....and more
```

File: include/utils/cutensornet.hpp
[the remark this answers is not in the document]
Comment (j9263178) [reply include/utils/cutensornet.hpp:1]: Now we don't store `cutensornet` object in `Network`, instead we create `cutensornet` when we need it during `setOrder()` or `Launch()`.

**Comment by j9263178** [comment]
> There are `new-delete-type-mismatch` in the `.FromString()` step and will cause the correspoding tests failed, running `NetworkTest.Network_dense_FromString` gives output:
> 
> ```
> ==1170185==ERROR: AddressSanitizer: new-delete-type-mismatch on 0x6140000c6c40 in thread T0:
>   object passed to delete has wrong type:
>   size of the allocated type:   424 bytes;
>   size of the deallocated type: 440 bytes.
>     #0 0x7f0eec4c0d3f in operator delete(void*, unsigned long) ../../../../libsanitizer/asan/asan_new_delete.cpp:172
>     #1 0x5627fcf89156 in cytnx::Network_base::~Network_base() /home/j9263178/Cytnx/include/Network.hpp:138
>     #2 0x5627fcb225e4 in void boost::checked_delete<cytnx::Network_base const>(cytnx::Network_base const*) /home/j9263178/anaconda3/envs/cytnx_cuda/include/boost/core/checked_delete.hpp:36
>     #3 0x5627fcb225e4 in cytnx::intrusive_ptr_release(cytnx::intrusive_ptr_base<cytnx::Network_base> const*) /home/j9263178/Cytnx/include/intrusive_ptr_base.hpp:47
>     #4 0x5627fcb225e4 in boost::intrusive_ptr<cytnx::Network_base>::~intrusive_ptr() /home/j9263178/anaconda3/envs/cytnx_cuda/include/boost/smart_ptr/intrusive_ptr.hpp:98
>     #5 0x5627fcb225e4 in boost::intrusive_ptr<cytnx::Network_base>::operator=(boost::intrusive_ptr<cytnx::Network_base> const&) /home/j9263178/anaconda3/envs/cytnx_cuda/include/boost/smart_ptr/intrusive_ptr.hpp:154
>     #6 0x5627fcb225e4 in cytnx::Network::FromString(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, int const&) /home/j9263178/Cytnx/include/Network.hpp:358
>     #7 0x5627fcb30ad0 in NetworkTest_Network_dense_FromString_Test::TestBody() /home/j9263178/Cytnx/tests/Network_test.cpp:23
>     #8 0x5627fcf348ce in void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /home/j9263178/Cytnx/build/_deps/googletest-src/googletest/src/gtest.cc:2589
>     #9 0x5627fcf348ce in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /home/j9263178/Cytnx/build/_deps/googletest-src/googletest/src/gtest.cc:2625
>     #10 0x5627fcf36161 in testing::Test::Run() /home/j9263178/Cytnx/build/_deps/googletest-src/googletest/src/gtest.cc:2664
> ....and more
> ```
> 
> And it is weird that deallocated size is larger than the allocated one.

To solve this I change `target_compile_definitions(cytnx PRIVATE UNI_GPU)` to `target_compile_definitions(cytnx PUBLIC UNI_GPU)`

**Comment by kaihsin** [comment]
whats the status of this PR now? does all the issues segfault resolved?

**Comment by j9263178** [comment]
> whats the status of this PR now? does all the issues segfault resolved?

Yes, I haven't encountered more issues and all the gpu and cpu network tests passed.

**Comment by kaihsin** [comment]
ok, then let me go through it and we should merge asap

**Review by kaihsin — APPROVED** [review]

**Comment by kaihsin** [comment]
please resolve the conflict by merge dev-master into this PR or rebase

**Comment by j9263178** [comment]
> please resolve the conflict by merge dev-master into this PR or rebase

I've merged dev-master into this PR.

**Pull request merged**